### PR TITLE
Enable precomputed table only if L2 quantizer and by_residual is true

### PR DESCRIPTION
**Pull request by chasingegg**
Related issue #2658

### CODE REVIEW

File: faiss/IndexIVFPQ.cpp [hunk 1]
```diff
@@ -422,6 +422,7 @@ void initialize_IVFPQ_precomputed_table(
         const Index* quantizer,
         const ProductQuantizer& pq,
         AlignedTable<float>& precomputed_table,
+        bool& by_residual,
```
Comment (mdouze) [inline faiss/IndexIVFPQ.cpp:425]: please pass in by value not reference

Comment (chasingegg) [reply faiss/IndexIVFPQ.cpp]: Updated.

File: faiss/IndexIVFPQ.cpp [hunk 2]
```diff
@@ -433,10 +434,10 @@ void initialize_IVFPQ_precomputed_table(
     }
 
     if (use_precomputed_table == 0) { // then choose the type of table
-        if (quantizer->metric_type == METRIC_INNER_PRODUCT) {
+        if (!(quantizer->metric_type == METRIC_L2 && by_residual)) {
             if (verbose) {
                 printf("IndexIVFPQ::precompute_table: precomputed "
-                       "tables not needed for inner product quantizers\n");
+                       "tables not needed only if L2 quantizer and by_residual is enabled\n");
```
Comment (mdouze) [inline faiss/IndexIVFPQ.cpp:440]: please change this to " tables needed only for  L2 metric and by_residual is enabled" and make sure the line is < 80 characters.

Comment (chasingegg) [reply faiss/IndexIVFPQ.cpp]: Done.

**Pull request closed without merging**